### ISSUE TITLE
LSP-like features for `.norg` file editing

### ISSUE THREAD

**Issue by vhyrro** [nvim-neorg/neorg#513]
### Discussed in https://github.com/nvim-neorg/neorg/discussions/512

<div type='discussions-op-text'>

<sup>Originally posted by **Aceticia** July 11, 2022</sup>
I'm writing my personal notes, journal, GTD, and wiki with neorg. With the telescope integration, things are seamless - I can search for content very effectively and link files. However, there are still some functionalities that are somewhat missing in maintaining a large scale wiki. Functions such as dead link detection, completion-based linking, previewing, links and backlinks, and renaming provided by markdown language servers such as [marksman](https://github.com/artempyanykh/marksman) would be a great compliment to the current, growing functionalities of neorg, for the people who take and maintain a lot of notes.</div>

This issue is mostly a reference point for future me on what to implement should I attempt the task of making a Neorg LSP. By all means if there's someone out there with more experience and they want to help out go ahead!

**Comment by Aceticia** [comment]
A more formal list of LSP features I can think of (feel free to point out what's missing/unfeasible):

**Diagnostics:**
- Broken link detection
- Incomplete blocks (e.g. meta data, code, ...)
- Overdue tasks

**Completion:**
- Link to file
- Link to header of a file
- Categories

**Hover:**
- Linked file/header preview

**Comment by danymat** [comment]
**GTD:**
- Unclassified tasks and projects detection

**Comment by Aceticia** [comment]
> I always wanted something like a potential link detection or proactive link completion. Not sure how to call it, so I'll try to give an example.
> 
> Imagine you have a file where you described what Foo is, like
> 
> ```
> * Foo
>   Foo is blah blah
> ** Foo use cases
>    You will use Foo when you want to blah
> ```
> 
> And then you're brainstorming ideas for something else:
> 
> ```
> * Bar
>   Possibly can be achieved with
>   - Something something
>   - Maybe can use Foo with blah blah
> ```
> 
> It will be very nice if Foo on the last line produced a completion/code action/code lense/whatever to turn `Foo` in the text into a link to that `Foo` header where you previously described what Foo is.
> 
> Could be very useful for wiki, knowledge base and zettelkasten workflows.

I'm actually thinking about how to make something like this. I know `obsidian.md` uses exact word match to suggest links, but I wanted to do some more fuzzy matching either based on word/sentence embeddings or n-gram models. 

That said, these are all quite high capacity computation, so likely not fast enough to run with LSP on each text insert. Running a command that detects potential links is a bit more feasible, in my opinion...but I might be wrong. I usually work with python NLP packages, and this is usually the speed I expect. Someone please correct me if you can run this with rust crates at text insert speed, that would be really cool.

**Comment by Aceticia** [comment]
True. Now that I think about it these might serve two slightly different purposes. The more accurate and faster algorithm for suggesting exact links, while the more fuzzy one is probably more appropriately used as a note exploration tool, like [noduslab](https://noduslabs.com)?

**Comment by aareman** [comment]
Providing an adapter to null-ls might be the easiest way to leverage existing neorg code and provide LSP/linting/formatting in an async way. (Just throwing this out there, I'm no expert at null-ls)

Thanks @vhyro and crew for this really cool plugin!
________________________________

**Comment by vhyrro** [comment]
@aareman i don't think using null-ls would be a good idea. Making a generic LSP would nicely help with Neorg adoption outside of Neovim, which is what we want, right? :)

**Comment by vhyrro** [comment]
Speaking of which, I've started work on an official LSP implementation here: https://github.com/nvim-neorg/neorg-lsp!

I've only been working on it for a day, so don't expect it to do anything useful :p, but I'm letting y'all know about its existence so people can contribute and so that you're aware that work is being done on it :)

**Comment by AlphabetsAlphabets** [comment]
I know I'm late to the party but something I'd like is to be able to update links when a file changes. In Obsidian, when you make a link to a file and change the file's name. All links get updated to match the file getting renamed. In Neorg that doesn't happen and when stuff breaks, I have to change a lot of things manually.

Having that would be very nice. The issue can be mitigated by just using anchors all over the place, so you can just change the path in one spot in a single file. But, it still has the issue of if the anchors are defined in different files.
